README.md
[![神策数据](https://opensource.sensorsdata.cn/wp-content/uploads/logo.png "神策数据")](https://www.sensorsdata.cn/)
<br><br>

[![License](https://img.shields.io/github/license/sensorsdata/sa-sdk-alipaymini.svg)](https://github.com/sensorsdata/sa-sdk-alipaymini/blob/master/LICENSE)
[![GitHub release](https://img.shields.io/github/release/sensorsdata/sa-sdk-alipaymini.svg)](https://github.com/sensorsdata/sa-sdk-alipaymini/releases)

# sa-sdk-alipaymini
## 神策简介

[**神策数据**](https://www.sensorsdata.cn/)
（Sensors Data），隶属于神策网络科技（北京）有限公司，是一家专业的大数据分析服务公司，大数据分析行业开拓者，为客户提供深度用户行为分析平台、以及专业的咨询服务和行业解决方案，致力于帮助客户实现数据驱动。神策数据立足大数据及用户行为分析的技术与实践前沿，业务现已覆盖以互联网、金融、零售快消、高科技、制造等为代表的十多个主要行业、并可支持企业多个职能部门。公司总部在北京，并在上海、深圳、合肥、武汉等地拥有本地化的服务团队，覆盖东区及南区市场；公司拥有专业的服务团队，为客户提供一对一的客户服务。公司在大数据领域积累的核心关键技术，包括在海量数据采集、存储、清洗、分析挖掘、可视化、智能应用、安全与隐私保护等领域。 [**More**](https://www.sensorsdata.cn/about/aboutus.html)

## 使用方法
使用方法请参考文档  https://www.sensorsdata.cn/manual/alipaymini_sdk.html

注意 SDK 可能不完全向前兼容，请查看版本更新说明 [CHANGELOG.md](CHANGELOG.md)。如果有说明不兼容的话，需要升级神策分析对应的版本。
请根据需要前往 [Releases](https://github.com/sensorsdata/sa-sdk-alipaymini/releases) 里下载对应的文件


## 版本更新记录

请参见 [CHANGELOG.md](CHANGELOG.md)

## 招聘信息
招聘小程序 SDK 开发工程师，小程序 SDK 技术顾问，简历发送到邮箱 (shengyonggen@sensorsdata.cn)<br>
详情可访问官网https://www.sensorsdata.cn/about/joinus.html

## 讨论

| 扫码加入神策数据开源社区 QQ 群<br>群号：785122381 | 扫码加入神策数据开源社区微信群 |
| ------ | ------ |
|![ QQ 讨论群](https://opensource.sensorsdata.cn/wp-content/uploads/ContentCommonPic_1.png) | ![ 微信讨论群 ](https://opensource.sensorsdata.cn/wp-content/uploads/ContentCommonPic_2.png) |

## 公众号

| 扫码关注<br>神策数据开源社区 | 扫码关注<br>神策数据开源社区服务号 |
| ------ | ------ |
|![ 微信订阅号 ](https://opensource.sensorsdata.cn/wp-content/uploads/ContentCommonPic_3.png) | ![ 微信服务号 ](https://opensource.sensorsdata.cn/wp-content/uploads/ContentCommonPic_4.png) |


## 新书推荐

| 《数据驱动：从方法到实践》 | 《Android 全埋点解决方案》 | 《iOS 全埋点解决方案》
| ------ | ------ | ------ |
| [![《数据驱动：从方法到实践》](https://opensource.sensorsdata.cn/wp-content/uploads/data_driven_book_1.jpg)](https://item.jd.com/12322322.html) | [![《Android 全埋点解决方案》](https://opensource.sensorsdata.cn/wp-content/uploads/Android-全埋点thumbnail_1.png)](https://item.jd.com/12574672.html) | [![《iOS 全埋点解决方案》](https://opensource.sensorsdata.cn/wp-content/uploads/iOS-全埋点thumbnail_1.png)](https://item.jd.com/12867068.html)

## License

Copyright 2015－2020 Sensors Data Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
## 1.2.1 (2022-4-22)
1. 新增
    - 支持 `Component` 组件点击事件采集

## 1.2.0 (2022-3-18)
1. 新增
    - 新增支持 A/B 功能 

## 1.1.9 (2022-2-14)
1. 修复
    - 修复 `getSystemInfo` 初始化时重复调用的问题

## 1.1.8 (2022-2-9)
1. 新增
    - 支付宝小程序 SDK 支持钉钉小程序

## 1.1.7 (2022-1-18)
1. 新增
    - 新增页面停留时长采集功能

## 1.1.6 (2021-9-15)
1. 优化
    - 修改初始化配置参数 `show_log` 默认值为 `fasle`
    - 修改初始化配置参数 `max_string_length` 默认值为 `500`
    - `$SignUp` 事件去除 `$is_first_day` 属性

## 1.1.5 (2021-8-31)
1. 修复
    - 修复自定义版本开发者工具真机调试网络请求失败的问题
    - 优化批量发送 `300` 条限制逻辑

## 1.1.4 (2021-7-16)
1. 新增
    - `$MPLaunch` 和 `$MPShow` 事件新增预置属性 `$url_query`;
    - 支持对特定控件不采集全埋点点击事件;
    - 新增获取预置属性的接口;
    - 新增获取数据接收地址的接口;

## 1.1.3 (2021-5-6)
1. 新增
    - 支持注册动态公共属性;

## 1.1.2 (2021-4-8)
1. 新增
    - 支持批量发送功能;

## 1.1.1 (2021-3-26)
1. 修复
    - 修复支付宝小程序基础库版本为 `2.0` 时，全埋点不采集 `$MPViewScreen` 事件问题;

## 1.1.0 (2021-3-23)
1. 新增
    - 全埋点预置事件支持自定义属性;

## 1.0.21 (2021-3-13)
1. 新增
    - 增加 `_flush_time` 和 `_track_id` 属性；

## 1.0.20 (2021-1-19)
1. 新增
    - 全埋点支持 `Taro` 框架
2. 优化
    - 优化场景值映射，SA 需升级到 `2.1.5107` 及以上版本，SDG 需升级到 `0.7.5159` 及以上版本；

## 1.0.18 (2020-11-16)
1. 新增
    - 支持渠道短链功能；

## 1.0.17 (2020-10-30)
1. 新增
    - 新增经纬度采集；

## 1.0.16 (2020-09-16)
1. 新增
    - 新增获取匿名 ID 接口；

## 1.0.15 (2020-09-12)
1. 优化
    - 数据接收地址优化；
    - 增加最近一次启动场景；

## 1.0.14 (2020-08-05)
1. 修复
    - 修复请求后端超时问题；

## 1.0.13 (2020-06-30)
1. 新增
    - 新增所有事件都有的预置属性 `$app_id` 和 `$timezone_offset`；

## 1.0.12 (2020-04-23)
1. 修复
    - 修复页面渠道参数解析时增加空对象参数判断；

## 1.0.11 (2020-04-13)
1. 优化
    - 优化渠道采集功能逻辑；
2. 新增
    - 新增控制台 `log` 日志显示开启/关闭功能；
    - 新增新用户访问时向用户表中写 `$first_visit_time` 预置属性；

## 1.0.10 (2020-03-16)
1. 新增
    - 新增全埋点版 `SDK` 新增预置事件可配置功能，默认开启；
2. 修复
    - 修复已知 `bug`；

## 1.0.9 (2020-03-05)
1. 新增
    - 新增 `logout()` 接口；
    - 新增 `init()` 接口，用来控制数据的发送时机；

## 1.0.8 (2020-02-27)
1. 修复
    - 修复事件处理函数中调用其他有返回值的函数时，返回值被修改为 `undefined` 的问题；

## 1.0.7 (2020-02-26)
1. 新增
    - 新增 `sensorsdata.custom.min.js` 文件，去除了 `sensorsdata_conf.js` 文件，初始化时使用 `setPara()` 方法初始化参数；
2. 修复
    - 修复点击事件处理函数中调用其他函数时传递对象参数报错问题；

## 1.0.6 (2020-02-17)
1. 新增
    - 支持全埋点版本，自动采集 `$MPLaunch`（小程序启动)、`$MPShow`（小程序显示）、`$MPHide`（小程序进入后台）、`$MPViewScreen`（小程序页面浏览）、`$MPClick`（小程序元素点击）五个预置事件；

## 1.0.5
1. 优化
    - 统一格式化所有小程序的预置属性 `$os` 的取值，从系统默认的 `ios` 改成了 `iOS` ，`android` 改成了 `Android`

## 1.0.4
1. 优化
    - 设置网络请求时的 `dataType` 为 `text`

## 1.0.3
1. 修复
    - 修正了扫描普通二维码打开小程序代码报错的问题

## 1.0.2
1. 优化
    - 更新发送请求的方式，兼容 `my.request` 和 `my.httpRequest`

## 1.0.1
* 增加了预置事件 $MPLaunch, $MPShow, $MPHide


sensorsdata.js
/**
 * @fileoverview sensors analytic miniprogram sdk
 * @author shengyonggen@sensorsdata.cn
 */
import conf from './sensorsdata_conf.js';

var sa = {};
var _ = {};

sa.para = conf;
sa.para.max_string_length = sa.para.max_string_length || 200;

sa._queue = [];
// 是否已经获取到系统信息
sa.getSystemInfoComplete = false;

var ArrayProto = Array.prototype,
  FuncProto = Function.prototype,
  ObjProto = Object.prototype,
  slice = ArrayProto.slice,
  toString = ObjProto.toString,
  hasOwnProperty = ObjProto.hasOwnProperty,
  LIB_VERSION = '0.2',
  LIB_NAME = 'AlipayMini';

var source_channel_standard = 'utm_source utm_medium utm_campaign utm_content utm_term';

var sa_referrer = '直接打开';
sa.lib_version = LIB_VERSION;

var logger = typeof logger === 'object' ? logger : {};
logger.info = function () {
  if (typeof console === 'object' && console.log) {
    try {
      return console.log.apply(console, arguments);
    } catch (e) {
      console.log(arguments[0]);
    }
  }
};

(function () {
  var nativeBind = FuncProto.bind,
    nativeForEach = ArrayProto.forEach,
    nativeIndexOf = ArrayProto.indexOf,
    nativeIsArray = Array.isArray,
    breaker = {};

  var each = _.each = function (obj, iterator, context) {
    if (obj == null) {
      return false;
    }
    if (nativeForEach && obj.forEach === nativeForEach) {
      obj.forEach(iterator, context);
    } else if (obj.length === +obj.length) {
      for (var i = 0,
        l = obj.length; i < l; i++) {
        if (i in obj && iterator.call(context, obj[i], i, obj) === breaker) {
          return false;
        }
      }
    } else {
      for (var key in obj) {
        if (hasOwnProperty.call(obj, key)) {
          if (iterator.call(context, obj[key], key, obj) === breaker) {
            return false;
          }
        }
      }
    }
  };

  _.logger = logger;
  // 普通的extend，不能到二级
  _.extend = function (obj) {
    each(slice.call(arguments, 1), function (source) {
      for (var prop in source) {
        if (source[prop] !==
          void 0) {
          obj[prop] = source[prop];
        }
      }
    });
    return obj;
  };
  // 允许二级的extend
  _.extend2Lev = function (obj) {
    each(slice.call(arguments, 1), function (source) {
      for (var prop in source) {
        if (source[prop] !==
          void 0) {
          if (_.isObject(source[prop]) && _.isObject(obj[prop])) {
            _.extend(obj[prop], source[prop]);
          } else {
            obj[prop] = source[prop];
          }
        }
      }
    });
    return obj;
  };
  // 如果已经有的属性不覆盖,如果没有的属性加进来
  _.coverExtend = function (obj) {
    each(slice.call(arguments, 1), function (source) {
      for (var prop in source) {
        if (source[prop] !==
          void 0 && obj[prop] ===
          void 0) {
          obj[prop] = source[prop];
        }
      }
    });
    return obj;
  };

  _.isArray = nativeIsArray ||
    function (obj) {
      return toString.call(obj) === '[object Array]';
    };

  _.isFunction = function (f) {
    try {
      return /^\s*\bfunction\b/.test(f);
    } catch (x) {
      return false;
    }
  };

  _.isArguments = function (obj) {
    return !!(obj && hasOwnProperty.call(obj, 'callee'));
  };

  _.toArray = function (iterable) {
    if (!iterable) {
      return [];
    }
    if (iterable.toArray) {
      return iterable.toArray();
    }
    if (_.isArray(iterable)) {
      return slice.call(iterable);
    }
    if (_.isArguments(iterable)) {
      return slice.call(iterable);
    }
    return _.values(iterable);
  };

  _.values = function (obj) {
    var results = [];
    if (obj == null) {
      return results;
    }
    each(obj, function (value) {
      results[results.length] = value;
    });
    return results;
  };

  _.include = function (obj, target) {
    var found = false;
    if (obj == null) {
      return found;
    }
    if (nativeIndexOf && obj.indexOf === nativeIndexOf) {
      return obj.indexOf(target) != -1;
    }
    each(obj, function (value) {
      if (found || (found = (value === target))) {
        return breaker;
      }
    });
    return found;
  };

})();

_.trim = function (str) {
  return str.replace(/^[\s\uFEFF\xA0]+|[\s\uFEFF\xA0]+$/g, '');
};

_.isObject = function (obj) {
  return (toString.call(obj) == '[object Object]') && (obj != null);
};

_.isEmptyObject = function (obj) {
  if (_.isObject(obj)) {
    for (var key in obj) {
      if (hasOwnProperty.call(obj, key)) {
        return false;
      }
    }
    return true;
  }
  return false;
};

_.isUndefined = function (obj) {
  return obj ===
    void 0;
};

_.isString = function (obj) {
  return toString.call(obj) == '[object String]';
};

_.isDate = function (obj) {
  return toString.call(obj) == '[object Date]';
};

_.isBoolean = function (obj) {
  return toString.call(obj) == '[object Boolean]';
};

_.isNumber = function (obj) {
  return (toString.call(obj) == '[object Number]' && /[\d\.]+/.test(String(obj)));
};

_.isJSONString = function (str) {
  try {
    JSON.parse(str);
  } catch (e) {
    return false;
  }
  return true;
};
// gbk等编码decode会异常
_.decodeURIComponent = function (val) {
  var result = '';
  try {
    result = decodeURIComponent(val);
  } catch (e) {
    result = val;
  };
  return result;
};

_.encodeDates = function (obj) {
  _.each(obj, function (v, k) {
    if (_.isDate(v)) {
      obj[k] = _.formatDate(v);
    } else if (_.isObject(v)) {
      obj[k] = _.encodeDates(v);
      // recurse
    }
  });
  return obj;
};

_.formatDate = function (d) {
  function pad(n) {
    return n < 10 ? '0' + n : n;
  }

  return d.getFullYear() + '-' + pad(d.getMonth() + 1) + '-' + pad(d.getDate()) + ' ' + pad(d.getHours()) + ':' + pad(d.getMinutes()) + ':' + pad(d.getSeconds()) + '.' + pad(d.getMilliseconds());
};

// 把日期格式全部转化成日期字符串
_.searchObjDate = function (o) {
  if (_.isObject(o)) {
    _.each(o, function (a, b) {
      if (_.isObject(a)) {
        _.searchObjDate(o[b]);
      } else {
        if (_.isDate(a)) {
          o[b] = _.formatDate(a);
        }
      }
    });
  }
};
// 把字符串格式数据限制字符串长度
_.formatString = function (str) {
  if (str.length > sa.para.max_string_length) {
    logger.info('字符串长度超过限制，已经做截取--' + str);
    return str.slice(0, sa.para.max_string_length);
  } else {
    return str;
  }
};

// 把字符串格式数据限制字符串长度
_.searchObjString = function (o) {
  if (_.isObject(o)) {
    _.each(o, function (a, b) {
      if (_.isObject(a)) {
        _.searchObjString(o[b]);
      } else {
        if (_.isString(a)) {
          o[b] = _.formatString(a);
        }
      }
    });
  }
};

// 数组去重复
_.unique = function (ar) {
  var temp,
    n = [],
    o = {};
  for (var i = 0; i < ar.length; i++) {
    temp = ar[i];
    if (!(temp in o)) {
      o[temp] = true;
      n.push(temp);
    }
  }
  return n;
};

// 只能是sensors满足的数据格式
_.strip_sa_properties = function (p) {
  if (!_.isObject(p)) {
    return p;
  }
  _.each(p, function (v, k) {
    // 如果是数组，把值自动转换成string
    if (_.isArray(v)) {
      var temp = [];
      _.each(v, function (arrv) {
        if (_.isString(arrv)) {
          temp.push(arrv);
        } else {
          logger.info('您的数据-', v, '的数组里的值必须是字符串,已经将其删除');
        }
      });
      if (temp.length !== 0) {
        p[k] = temp;
      } else {
        delete p[k];
        logger.info('已经删除空的数组');
      }
    }
    // 只能是字符串，数字，日期,布尔，数组
    if (!(_.isString(v) || _.isNumber(v) || _.isDate(v) || _.isBoolean(v) || _.isArray(v))) {
      logger.info('您的数据-', v, '-格式不满足要求，我们已经将其删除');
      delete p[k];
    }
  });
  return p;
};

// 去掉undefined和null
_.strip_empty_properties = function (p) {
  var ret = {};
  _.each(p, function (v, k) {
    if (v != null) {
      ret[k] = v;
    }
  });
  return ret;
};

_.utf8Encode = function (string) {
  string = (string + '').replace(/\r\n/g, '\n').replace(/\r/g, '\n');

  var utftext = '',
    start,
    end;
  var stringl = 0,
    n;

  start = end = 0;
  stringl = string.length;

  for (n = 0; n < stringl; n++) {
    var c1 = string.charCodeAt(n);
    var enc = null;

    if (c1 < 128) {
      end++;
    } else if ((c1 > 127) && (c1 < 2048)) {
      enc = String.fromCharCode((c1 >> 6) | 192, (c1 & 63) | 128);
    } else {
      enc = String.fromCharCode((c1 >> 12) | 224, ((c1 >> 6) & 63) | 128, (c1 & 63) | 128);
    }
    if (enc !== null) {
      if (end > start) {
        utftext += string.substring(start, end);
      }
      utftext += enc;
      start = end = n + 1;
    }
  }

  if (end > start) {
    utftext += string.substring(start, string.length);
  }

  return utftext;
};

_.base64Encode = function (data) {
  var b64 = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=';
  var o1,
    o2,
    o3,
    h1,
    h2,
    h3,
    h4,
    bits,
    i = 0,
    ac = 0,
    enc = '',
    tmp_arr = [];
  if (!data) {
    return data;
  }
  data = _.utf8Encode(data);
  do {
    o1 = data.charCodeAt(i++);
    o2 = data.charCodeAt(i++);
    o3 = data.charCodeAt(i++);

    bits = o1 << 16 | o2 << 8 | o3;

    h1 = bits >> 18 & 0x3f;
    h2 = bits >> 12 & 0x3f;
    h3 = bits >> 6 & 0x3f;
    h4 = bits & 0x3f;
    tmp_arr[ac++] = b64.charAt(h1) + b64.charAt(h2) + b64.charAt(h3) + b64.charAt(h4);
  } while (i < data.length);

  enc = tmp_arr.join('');

  switch (data.length % 3) {
    case 1:
      enc = enc.slice(0, -2) + '==';
      break;
    case 2:
      enc = enc.slice(0, -1) + '=';
      break;
  }

  return enc;
};

_.info = {
  currentProps:{},
  properties: {
    $lib: LIB_NAME,
    $lib_version: String(LIB_VERSION),
    $user_agent: 'SensorsAnalytics AlipayMini SDK'
  },
  getSystem: function () {
    var e = this.properties;
    var that = this;
    // 小程序bug，某些版本没有complete参数。
    var systemNotCalled = true;

    function getSystemInfo() {
      if(systemNotCalled){
        systemNotCalled = true;
        my.getSystemInfo({
          success : function (t) {
            e.$model = t["model"];
            e.$screen_width = Number(t["screenWidth"]);
            e.$screen_height = Number(t["screenHeight"]);
            e.$os = t["platform"];
            //e.$os_version = t["version"];
            e.$manufacturer = t["brand"];
            that.setStatusComplete();
          },
          complete : that.setStatusComplete
        });
      }
    }

    function getNetwork() {

      my.getNetworkType({
        success : function (t) {
          e.$network_type = t["networkType"];
          getSystemInfo();
        },
        complete : function(){
          getSystemInfo();
        }
      });

    }

    getNetwork();
  },
  setStatusComplete: function () {
    if(sa.getSystemInfoComplete){
      return false;
    }
    sa.getSystemInfoComplete = true;
    if (sa._queue.length > 0) {
      _.each(sa._queue, function (content) {
        sa.prepareData.apply(sa, slice.call(content));
      });
      sa._queue = [];
    }
  }
};

sa._ = _;

sa.prepareData = function (p, callback) {
   if (!sa.getSystemInfoComplete) {
     sa._queue.push(arguments);
     return false;
   }

  var data = {
    distinct_id: this.store.getDistinctId(),
    lib: {
      $lib: LIB_NAME,
      $lib_method: 'code',
      $lib_version: String(LIB_VERSION)
    },
    properties: {}
  };
  _.extend(data, p);

  // 合并properties里的属性
  if (_.isObject(p.properties) && !_.isEmptyObject(p.properties)) {
    _.extend(data.properties, p.properties);
  }

  // profile时不传公用属性
  if (!p.type || p.type.slice(0, 7) !== 'profile') {
    // 传入的属性 > 当前页面的属性 > session的属性 > cookie的属性 > 预定义属性
    data.properties = _.extend({}, _.info.properties, sa.store.getProps(),  _.info.currentProps, data.properties);
    // 判断是否是首日访问，果子说要做
    if (typeof sa.store._state === 'object' && typeof sa.store._state.first_visit_day_time === 'number' && sa.store._state.first_visit_day_time > (new Date()).getTime()) {
      data.properties.$is_first_day = true;
    } else {
      data.properties.$is_first_day = false;
    }
  }
  // 如果$time是传入的就用，否则使用服务端时间
  if (data.properties.$time && _.isDate(data.properties.$time)) {
    data.time = data.properties.$time * 1;
    delete data.properties.$time;
  } else {

      data.time = (new Date()) * 1;

  }


  _.searchObjDate(data);
  _.searchObjString(data);
  sa.send(data, callback);
};

sa.store = {
  getUUID: function () {
    return "" + Date.now() + '-' + Math.floor(1e7 * Math.random()) + '-' + Math.random().toString(16).replace('.', '') + '-' + String(Math.random() * 31242).replace('.', '').slice(0, 8);

  },
  setStorage: function () {

  },
  getStorage: function () {
    return my.getStorageSync({key: 'sensorsdata2015_zfb'}) || {};
  },
  _state: {},
  toState: function (state) {
    if (typeof state === 'object') {
      if (state.distinct_id) {
        this._state = state;
      } else {
        this.set('distinct_id', this.getUUID());
      }
    } else {
      this.set('distinct_id', this.getUUID());
    }
  },
  getFirstId: function () {
    return this._state.first_id;
  },
  getDistinctId: function () {
    return this._state.distinct_id;
  },
  getProps: function () {
    return this._state.props || {};
  },
  setProps: function (newp, isCover) {
    var props = this._state.props || {};
    if (!isCover) {
      _.extend(props, newp);
      this.set('props', props);
    } else {
      this.set('props', newp);
    }
  },
  set: function (name, value) {
    var obj = {};
    if (typeof name === 'string') {
      obj[name] = value;
    } else if (typeof name === 'object') {
      obj = name;
    }
    this._state = this._state || {};
    for (var i in obj) {
      this._state[i] = obj[i];
    }
    this.save();
  },
  change: function (name, value) {
    this._state[name] = value;
  },
  save: function () {
    my.setStorageSync({
      key: 'sensorsdata2015_zfb',
      data: this._state
    });
    
  },
  init: function () {
    var info = this.getStorage().data;
    if (info) {
      this.toState(info);
    } else {
      var time = (new Date());
      var visit_time = time.getTime();
      time.setHours(23);
      time.setMinutes(59);
      time.setSeconds(60);
      this.set({
        'distinct_id': this.getUUID(),
        'first_visit_time': visit_time,
        'first_visit_day_time': time.getTime()
      });
    }
  }
};

sa.setProfile = function (p, c) {
  sa.prepareData({
    type: 'profile_set',
    properties: p
  }, c);
};

sa.setOnceProfile = function (p, c) {
  sa.prepareData({
    type: 'profile_set_once',
    properties: p
  }, c);
};

sa.track = function (e, p, c) {
  this.prepareData({
    type: 'track',
    event: e,
    properties: p
  }, c);
};

sa.identify = function (id, isSave) {
  if (typeof id === 'number') {
    id = String(id);
  } else if (typeof id !== 'string') {
    return false;
  }
  var firstId = sa.store.getFirstId();
  if (isSave === true) {
    if (firstId) {
      sa.store.set('first_id', id);
    } else {
      sa.store.set('distinct_id', id);
    }
  } else {
    if (firstId) {
      sa.store.change('first_id', id);
    } else {
      sa.store.change('distinct_id', id);
    }
  }
};

sa.trackSignup = function (id, e, p, c) {

  sa.prepareData({
    original_id: sa.store.getFirstId() || sa.store.getDistinctId(),
    distinct_id: id,
    type: 'track_signup',
    event: e,
    properties: p
  }, c);
  sa.store.set('distinct_id', id);
};

sa.registerApp = function (obj) {
  if (_.isObject(obj) && !_.isEmptyObject(obj)) {
    _.info.currentProps = _.extend(_.info.currentProps, obj);
  }
};

/*
sa.register = function (obj) {
  if (_.isObject(obj) && !_.isEmptyObject(obj)) {
    sa.store.setProps(obj);
  }
};*/

sa.clearAllRegister = function () {
  sa.store.setProps({}, true);
};

sa.login = function (id) {
  var firstId = sa.store.getFirstId();
  var distinctId = sa.store.getDistinctId();
  if (id !== distinctId) {
    if (firstId) {
      sa.trackSignup(id, '$SignUp');
    } else {
      sa.store.set('first_id', distinctId);
      sa.trackSignup(id, '$SignUp');
    }
  }
};

sa.init = function () {
  this._.info.getSystem();

  this.store.init();
  if (_.isObject(this.para.register)) {
    _.info.properties = _.extend(_.info.properties, this.para.register);
  }

};

sa.send = function (t) {
  var url = '';
  t._nocache = (String(Math.random()) + String(Math.random()) + String(Math.random())).slice(2, 15);

  logger.info(t);

  t = JSON.stringify(t);

  if (sa.para.server_url.indexOf('?') !== -1) {
    url = sa.para.server_url + '&data=' + encodeURIComponent(_.base64Encode(t));
  } else {
    url = sa.para.server_url + '?data=' + encodeURIComponent(_.base64Encode(t));
  }

  var sendRequest = function () {
    my.httpRequest({
      "url": url,
      "method": "GET"
    });
  };
  sendRequest();
};


sa.init();



export default sa;